### PR TITLE
#2352 Number filter

### LINKED ISSUE #2352: Cold chain generic filters -- Number (range) filter
Part of Epic #2282 

~~Suggest using ":" to split upper and lower number in URL as per #2351~~ (Now using `"_"`)

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fixes #2352

# 👩🏻‍💻 What does this PR do? 
 

Generic number filter. Fairly simple, as the more complex functionality has been created in the earlier PRs.

I've also extended the main `<NumericTextInput>` component to take min, max and decimal limit props. I really think we can do away with the three other Number components (NonNegativeIntegerInput, NonNegativeNumberInput, and PositiveNumberInput) as they're basically duplicated code except for a couple of tiny things -- we can achieve the exact same thing with one component, these three additional optional props, and a single line of code to handle all three. Let me know if there's any good reason to keep them, otherwise I'll make a new PR to remove them based off this.

# 🧪 How has/should this change been tested? 

Again, not actually implemented in any real cases yet, so as before, just replace the `<FilterMenu />` code in `client/packages/system/src/Patient/ListView/Toolbar.tsx` with the following in order to test it out:

```tsx
<FilterMenu
	filters={[
	// Using the grouping feature to combine first and last name into a
	// single menu option
	{
	  type: 'group',
	  name: 'Name',
	  elements: [
		{
		  type: 'text',
		  name: t('label.first-name'),
		  urlParameter: 'firstName',
		  placeholder: t('placeholder.search-by-first-name'),
		},
		{
		  type: 'text',
		  name: t('label.last-name'),
		  urlParameter: 'lastName',
		  placeholder: t('placeholder.search-by-last-name'),
		},
	  ],
	},
	// A standalone number filter
	{
	  type: 'number',
	  name: 'Numeric value',
	  urlParameter: 'value',
	  minValue: 0,
	  maxValue: 100,
	},
	// Grouping two numbers together to form a "range"
	{
	  type: 'group',
	  name: 'Number range',
	  elements: [
		{
		  type: 'number',
		  name: 'Lower bound',
		  urlParameter: 'valueRange',
		  range: 'from',
		  decimalLimit: 2,
		},
		{
		  type: 'number',
		  name: 'Upper bound',
		  urlParameter: 'valueRange',
		  range: 'to',
		  decimalLimit: 2,
		},
	  ],
	},
	{
	  type: 'text',
	  name: t('label.patient-id'),
	  urlParameter: 'identifier',
	  placeholder: t('placeholder.search-by-identifier'),
	},
	// A standalone "Date" input
	{
	  type: 'date',
	  name: 'Date',
	  urlParameter: 'date',
	},
	// A standalone "Date/Time" input
	{
	  type: 'dateTime',
	  name: 'Date/Time',
	  urlParameter: 'dateTime',
	},
	// Grouping two date inputs to form a "range". We define which
	// element is which end of the range using the "range" property.
	// Note they both update the same urlParameter
	{
	  type: 'group',
	  name: 'Date Range',
	  elements: [
		{
		  type: 'date',
		  name: 'From date',
		  urlParameter: 'dateRange',
		  range: 'from',
		},
		{
		  type: 'date',
		  name: 'To date',
		  urlParameter: 'dateRange',
		  range: 'to',
		},
	  ],
	},
	// Same as above but using date/time
	{
	  type: 'group',
	  name: 'Date/Time Range',
	  elements: [
		{
		  type: 'dateTime',
		  name: 'From time',
		  urlParameter: 'dateTimeRange',
		  range: 'from',
		},
		{
		  type: 'dateTime',
		  name: 'To time',
		  urlParameter: 'dateTimeRange',
		  range: 'to',
		},
	  ],
	},
	{
	  type: 'enum',
	  name: t('label.gender'),
	  urlParameter: 'gender',
	  options: [
		{ label: 'Male', value: 'MALE' },
		{ label: 'Female', value: 'FEMALE' },
	  ],
	},
	]}
	/>
```

## 💌 Any notes for the reviewer?

See inline comments for further details.

### CODE REVIEW

File: client/packages/common/src/ui/components/inputs/TextInput/numeric/NumericTextInput.tsx [hunk 1]
```diff
@@ -22,7 +38,10 @@ export const NumericTextInput: FC<NumericTextInputProps> = React.forwardRef(
           return;
         }
         const parsed = Number(e.target.value);
-        if (!Number.isNaN(parsed) && !!onChange) onChange(parsed);
+        if (!Number.isNaN(parsed) && !!onChange)
+          onChange(
+            NumUtils.constrain(NumUtils.round(parsed, decimalLimit), min, max)
+          );
       }}
```
Comment (CarlosNZ) [inline client/packages/common/src/ui/components/inputs/TextInput/numeric/NumericTextInput.tsx:45]: This is all that is needed to incorporate all the functionality of the three other number input components, hence my desire to remove them :)

File: client/packages/common/src/ui/components/inputs/Filters/NumberFilter.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+  const handleChange = (newValue: number | undefined) => {
+    setValue(newValue);
+    debouncedOnChange(newValue);
+  };
+
+  return (
+    <NumericTextInput
+      label={name}
+      width={FILTER_WIDTH / 2}
```
[the remark this answers is not in the document]
Comment (roxy-dao) [reply client/packages/common/src/ui/components/inputs/Filters/NumberFilter.tsx:58]: Sounds good

File: client/packages/common/src/ui/components/inputs/TextInput/numeric/NumericTextInput.tsx (hunk 1 above)
Comment (roxy-dao) [reply client/packages/common/src/ui/components/inputs/TextInput/numeric/NumericTextInput.tsx:45]: https://github.com/openmsupply/open-msupply/issues/1440

**Review by roxy-dao — CHANGES_REQUESTED** [review]
Thanks Carl~ 

Everything looks great and works as stated. Code looks fine too. Just have one comment though... Could you put some validation in range so that it shows an error if the lower bound is higher than the upper bound?

**Review by roxy-dao — APPROVED** [review]

File: client/packages/common/src/ui/components/inputs/TextInput/numeric/NumericTextInput.tsx (hunk 1 above)
Comment (CarlosNZ) [reply client/packages/common/src/ui/components/inputs/TextInput/numeric/NumericTextInput.tsx:45]: > #1440

Nice -- I thought I'd encountered this before. Have added follow-up comment there to say that this is half-done now: https://github.com/openmsupply/open-msupply/pull/2381#discussion_r1366006870

**Pull request merged**